operator/controllers/actionsrunner/reconciler.go

```diff
@@ -0,0 +1,298 @@
+/*
+Copyright 2020 In Loco Tecnologia da Informação S.A.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package actionsrunner
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	inlocov1alpha1 "github.com/inloco/kube-actions/operator/api/v1alpha1"
+	"github.com/inloco/kube-actions/operator/controllers/actionsrunner/facades"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+var (
+	patchOpts = []client.PatchOption{
+		client.ForceOwnership,
+		client.FieldOwner("kube-actions"),
+	}
+
+	deleteOpts = []client.DeleteOption{
+		client.PropagationPolicy(metav1.DeletePropagationForeground),
+	}
+)
+
+// Reconciler reconciles a ActionsRunner object
+type Reconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+
+	source  chan event.GenericEvent
+	waiting sync.Map
+	wires   map[types.NamespacedName]*Wire
+}
+
+// +kubebuilder:rbac:groups=inloco.com.br,resources=actionsrunners,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=inloco.com.br,resources=actionsrunners/status,verbs=get;update;patch
+
+func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+	log := r.Log.WithValues("actionsrunner", req.NamespacedName)
+
+	var actionsRunner inlocov1alpha1.ActionsRunner
+	if err := r.Get(ctx, req.NamespacedName, &actionsRunner); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	var actionsRunnerJob inlocov1alpha1.ActionsRunnerJob
+	switch err := r.Get(ctx, req.NamespacedName, &actionsRunnerJob); {
+	case err == nil:
+		key := fmt.Sprintf("%s/%s", actionsRunnerJob.GetNamespace(), actionsRunnerJob.GetName())
+		if _, loaded := r.waiting.LoadOrStore(key, nil); !loaded {
```

Comment:
Not sure if this will remain here (as we may want to encapsulate it something close to the `CreateOfFetch` that I mentioned earlier), but I would write a comment here just to facilitate a first read of the code

```
// if the job existed already, atomically spawns a new background routine for waiting its completion
```

Reply:
I'll create another structure to encompass that